### PR TITLE
Add assert(predicate) syntax

### PR DESCRIPTION
**Pull request by TimWSpence**
Enables `IO.pure(1).assert(_ > 0)` syntax

### CODE REVIEW

**Comment by TimWSpence** [comment]
This was something that seemed useful to me earlier. Feel free to discard if you disagree 😅

File: core/src/main/scala/munit/CatsEffectAssertions.scala [hunk 1]
```diff
@@ -93,6 +93,30 @@ trait CatsEffectAssertions { self: Assertions =>
...
+    *   a value that will be printed in case the assertions fails
+    * @param pred
+    *   the predicate that must be satisfied
+    */
+  protected def assertIOPredicate[A](
+      obtained: IO[A]
+  )(pred: A => Boolean, clue: => Any = "predicate not satisfied")(implicit
+      loc: Location
+  ): IO[Unit] =
+    assertIO(obtained.map(pred), true, clue)
```
Comment (armanbilge) [inline core/src/main/scala/munit/CatsEffectAssertions.scala:118]: Lol, was going to say, isn't this basically a `map` 😛

Comment (danicheg) [reply core/src/main/scala/munit/CatsEffectAssertions.scala:118]: Agree with Arman on that
```scala
foo.map(_ > 42).assert
```
that seems even more concise IMO

Comment (TimWSpence) [reply core/src/main/scala/munit/CatsEffectAssertions.scala:118]: I thought it communicated intent more clearly. But yes, it is of course equivalent

Comment (TimWSpence) [reply core/src/main/scala/munit/CatsEffectAssertions.scala:118]: Not in this formulation but via syntax you have `foo.assert(_ > 42)` which is surely more concise?

Comment (danicheg) [reply core/src/main/scala/munit/CatsEffectAssertions.scala:118]: Ah, yes, my bad, I reacted to Arman's comment and didn't scroll the PR to the end 🤦🏻 
Anyway, can we leave the syntax methods only? Don't see evil to have them.

Comment (TimWSpence) [reply core/src/main/scala/munit/CatsEffectAssertions.scala:118]: Ah yeah, that makes more sense actually 👍

Comment (armanbilge) [reply core/src/main/scala/munit/CatsEffectAssertions.scala:118]: I also didn't scroll 😇

Comment (TimWSpence) [reply core/src/main/scala/munit/CatsEffectAssertions.scala]: OK done! I added `assertSyncIOBoolean` for consistency with the `IO` methods but can remove it if you prefer - it's not really very valuable either

File: core/src/main/scala/munit/CatsEffectAssertions.scala [hunk 2]
```diff
@@ -295,6 +316,25 @@ trait CatsEffectAssertions { self: Assertions =>
...
+      *
+      * The "clue" value can be used to give extra information about the failure in case the
+      * assertion fails.
+      *
+      * @param pred
+      *   the predicate that must be satisfied
+      * @param clue
+      *   a value that will be printed in case the assertions fails
+      */
+    def assert(pred: A => Boolean, clue: => Any = "predicate not satisfied")(implicit
```
Comment (armanbilge) [inline core/src/main/scala/munit/CatsEffectAssertions.scala:333]: I guess mixing the `pred` and `clue` into the same arg-list prevents doing something like:
```scala
ioa.assert { a =>
  foo(a)
}
```
I don't necessarily think that's a bad thing.

Comment (TimWSpence) [reply core/src/main/scala/munit/CatsEffectAssertions.scala:333]: Ah good point. That example works right? But 
```
ioa.assert { a =>
  foo(a), "some clue"
}
```
doesn't. Do you think I should curry it instead?

Comment (armanbilge) [reply core/src/main/scala/munit/CatsEffectAssertions.scala:333]: > That example works right?

Oh does it? My bad again 😂 so it does 😇 

https://scastie.scala-lang.org/eYBzk8oMRmKN00kxNKSg1g

> Do you think I should curry it instead?

I think that will force the extra set of parentheses always which seems even worse.

---

I think we can leave it. It seems like the most flexible/ergonomic is to do something like this:

```scala
ioa.map { a =>
  foo(a)
}.assert("some clue")
```

I think currently the `assert` syntax does not support adding a clue, but that's a separate problem.

Comment (TimWSpence) [reply core/src/main/scala/munit/CatsEffectAssertions.scala:333]: Oh ouch, hadn't realised you need empty parentheses if you just want to supply the default argument. Yeah, that's definitely worse!

Seems like people don't often care about manually supplying clues (as you say the existing `assert` doesn't support it) so I guess the current design is ok for the moment

**Comment by TimWSpence** [comment]
@armanbilge do I need to do anything with this? Or is it even wanted? 😂

**Review by armanbilge — APPROVED** [review]

**Comment by danicheg** [comment]
Thanks for this work, it's a lovely addition to the library.

**Comment by TimWSpence** [comment]
> Thanks for this work, it's a lovely addition to the library.

Thanks! :)

**Pull request merged**